velox/common/base/IndexedPriorityQueue.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <folly/container/F14Map.h>
+#include <folly/container/F14Set.h>
+#include <set>
+
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::velox {
+/// A priority queue with values of type 'T'. Each value has assigned priority
+/// on insertion which determines the value's position in the quue. It supports
+/// to update the priority of the existing values, which adjusts the value's
+/// position in the queue accordingly. Ties are broken by insertion order. If
+/// 'MaxQueue' is true, it is a max priority queue, otherwise a min priority
+/// queue.
+///
+/// NOTE: this class is not thread-safe.
+template <typename T, bool MaxQueue>
+class IndexedPriorityQueue {
+ public:
+  IndexedPriorityQueue() = default;
+
+  size_t size() const {
+    return queue_.size();
+  }
+
+  bool empty() const {
+    return queue_.empty();
+  }
+
+  bool addOrUpdate(T value, uint64_t priority) {
+    auto it = index_.find(value);
+    if (it != index_.end()) {
+      if (it->second->priority() == priority) {
+        return false;
+      }
+      queue_.erase(it->second.get());
+
+      it->second->updatePriority(priority);
+      queue_.insert(it->second.get());
+      return false;
+    }
+
+    auto newEntry = std::make_unique<Entry>(value, priority, generation_++);
+    queue_.insert(newEntry.get());
+    index_.emplace(value, std::move(newEntry));
+    return true;
+  }
+
+  std::optional<T> pop() {
+    VELOX_CHECK_EQ(queue_.size(), index_.size());
```

Comment:
This sanity check applies to all public APIs. If it is not performance sensitive we should be consistent and put the check also in addOrUpdate(). If not performance sensitive maybe we can just check   when the queue is popped empty.